cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -127,6 +124,69 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
 	return etcdClient, nil
 }
 
+// retrieveEtcdEndpoints returns the list of etcd endpoints.
+func retrieveEtcdEndpoints(client clientset.Interface) ([]string, error) {
+	etcdEndpoints, err := retrieveEtcdEndpointsFromPodAnnotationWithRetry(client)
+	if err != nil {
+		return retrieveEtcdEndpointsFromClusterStatus(client)
+	}
+	return etcdEndpoints, nil
+}
+
+// retrieveEtcdEndpointsFromPodAnnotation returns the list of endpoints as reported on etcd's pod annotations
+func retrieveEtcdEndpointsFromPodAnnotationWithRetry(client clientset.Interface) ([]string, error) {
+	etcdEndpoints := []string{}
+	var lastErr error
+	err := wait.ExponentialBackoff(clientsetretry.DefaultRetry, func() (bool, error) {
```

Comment:
the default backoff and retry are very short.
https://github.com/kubernetes/kubernetes/blob/8d247629c38228ac18513612024c527f58ed9c57/staging/src/k8s.io/client-go/util/retry/util.go#L26-L43

i think we should use `etcdBackoff` defined in this file for consistency even if this is retrying on client-go and the rest of the usages are for the etcd client and even if it's rather long.

Reply:
Thank you, I'll double check. I simply don't know what a reasonable default is right now. Note that we will forcefully have to timeout during the first `kubeadm upgrade apply`. In the case of this `etcd`, only the `apply` is affected, subsequent `kubeadm upgrade node` will find the first etcd pod after that.

Reply:
i don't remember us discussing this on the KEP. maybe we did.
wondering what would the best way be to handle this transitional upgrade.

maybe it makes sense to only use ClusterStatus for `upgrade apply` and not fetching the annotation for the 1.18 release.

this would require etcd.NewFromCluster() accepting another argument.

i did not think much about this.

Reply:
I'll think this one through a bit.